File: Heap-array-string-hash-pattern/0525-contiguous-array/solution.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n=nums.size(), res=0, sum=0;
        unordered_map <int,int> mp;

        for(int i=0;i<n;i++){
            nums[i]= (nums[i]==1)? 1: -1;
            sum+=nums[i];
            if(sum==0){
                res = max(res,i+1);
            }
            else if(mp.find(sum)!= mp.end()) {
                res= max(i-mp[sum], res);
            }
            else {
                mp[sum] = i;
            }
        }
        return res;
    }
};
      

File: Heap-array-string-hash-pattern/0525-contiguous-array/solution.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        first = {0:0}
        curr = 0
        ans = 0
        for i in range(len(nums)):
            curr+= 1 if nums[i]==1 else -1
            if curr in first:
                ans = max(ans, i+1-first[curr])
            else:
                first[curr] = i+1
        return ans
